File: clojure/ch01/ex1.35.clj
(ns sicp.ch01 (:use clojure.contrib.test-is))

; Calculate the golden ratio, phi.

(def tolerance 0.00001)

(defn abs [x] (if (> x 0) x (- x)))

(defn close-enough? [x y]
  (< (abs (- x y)) tolerance))

(defn average [x y]
  (/ (+ x y) 2.0))

(defn fixed-point [f first-guess]
  (defn fp [guess]
    (let [next-guess (f guess)]
      (cond (close-enough? guess next-guess) next-guess
            :else (fp next-guess))))
  (fp first-guess))

(deftest test-fixed-point
  (is (= 1.6180327868852458 (fixed-point (fn [x] (+ 1.0 (/ 1.0 x))) 1.0))))

(run-tests)

File: scala/ch01/ex1.35.scala
// Calculate the golden ratio, phi.

val tolerance = 0.00001

def abs(x: Double) = math.abs(x)

def closeEnough(x: Double, y: Double) = abs(x - y) < tolerance

def average(x: Double, y: Double) = (x + y) / 2.0

def fixedPoint(f: (Double) => Double, firstGuess: Double) = {
  def fp(guess: Double): Double = {
    val nextGuess = f(guess)
    if (closeEnough(guess, nextGuess))
      nextGuess
    else
      fp(nextGuess)
  }
  fp(firstGuess)
}

import org.scalatest._
import org.scalatest.matchers._

object fixedPointSpec extends Spec with ShouldMatchers {
  describe ("fixedPoint") {
    it ("should converge on the solution") {
      fixedPoint(x => 1.0 + (1.0 / x), 1.0) should equal (1.6180327868852458)
    }
  }
}

fixedPointSpec execute
